tools/benchmark/cmd/watch_latency.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/rand"
+	"os"
+	"time"
+
+	v3 "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/pkg/report"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/net/context"
+	"golang.org/x/time/rate"
+	"gopkg.in/cheggaaa/pb.v1"
+)
+
+// watchLatencyCmd represents the watch latency command
+var watchLatencyCmd = &cobra.Command{
+	Use:   "watch-latency",
+	Short: "Benchmark watch latency",
+	Long: `Benchmarks the latency for watches by measuring 
+	the latency between writing to a key and receiving the 
+	associated watch response.`,
+	Run: watchLatencyFunc,
+}
+
+var (
+	watchLTotal        int
+	watchLPutRate      int
+	watchLKeySize      int
+	watchLValueSize    int
+	watchLKeySpaceSize int
+	watchLSeqKey       bool
+)
+
+func init() {
+	RootCmd.AddCommand(watchLatencyCmd)
+	watchLatencyCmd.Flags().IntVar(&watchLTotal, "total", 10000, "Total number of watch responses.")
+	watchLatencyCmd.Flags().IntVar(&watchLPutRate, "put-rate", 100, "Number of keys to put per second")
+	watchLatencyCmd.Flags().IntVar(&watchLKeySize, "key-size", 32, "Key size of watch request")
+	watchLatencyCmd.Flags().IntVar(&watchLValueSize, "val-size", 32, "Val size of watch request")
+	watchLatencyCmd.Flags().IntVar(&watchLKeySpaceSize, "key-space-size", 1, "Maximum possible values")
+	watchLatencyCmd.Flags().BoolVar(&watchLSeqKey, "sequential-keys", false, "Use sequential values")
+}
+
+func watchLatencyFunc(cmd *cobra.Command, args []string) {
+	if watchLKeySpaceSize <= 0 {
+		fmt.Fprintf(os.Stderr, "expected positive --key-space-size, got (%v)", watchLKeySpaceSize)
+		os.Exit(1)
+	}
+
+	k := make([]byte, watchLValueSize)
+	if watchLSeqKey {
+		binary.PutVarint(k, int64(watchLKeySpaceSize))
+	} else {
+		binary.PutVarint(k, int64(rand.Intn(watchLKeySpaceSize)))
+	}
+
+	key := string(k)
+	value := string(mustRandBytes(watchLKeySize))
+
+	client := mustCreateConn()
+	stream := v3.NewWatcher(client)
+	wch := stream.Watch(context.TODO(), key)
+
+	bar = pb.New(watchLTotal)
+	bar.Format("Bom !")
+	bar.Start()
+
+	limiter := rate.NewLimiter(rate.Limit(watchLPutRate), watchLPutRate)
+	r := report.NewReport("%4.9f")
```

Comment:
@heyitsanthony actually i wanted the precision upto 9 decimal places, the newReport function is hardcoding it to %4.4f will try to add a function there to take the precision and change it as per your wish.